# python程序设计作业二 

作业内容：分析影响房价的因素

一、读取、查看数据：包括前5个元素，以及数据确实情况。

In [1]:
import pandas as pd
df = pd.read_csv('train.csv',index_col=0)

In [2]:
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig    ...     PoolArea PoolQC Fence  \
Id                                    ...                             
1          Lvl    AllPub    Inside    ...            0    NaN   NaN   
2          Lvl    AllPub       FR2    ...            0    NaN   NaN   
3          Lvl    AllPub    Inside    ...            0    NaN   NaN   
4          Lvl    AllPub    Corner    ...            0    NaN   NaN   
5          Lvl    AllPub       FR2    ...            0    NaN   NaN   

   MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                                         
1          NaN       0      2    2008        WD         Normal     208500  
2          NaN       0      5    2007        WD         Normal     181500  
3          NaN       0      9    2008        WD         Normal     223500  
4          NaN       0      2    2006        WD        Abnorml     140000  
5          NaN       0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

In [3]:
df.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1460.000000  1201.000000    1460.000000  1460.000000  1460.000000   
mean     56.897260    70.049958   10516.828082     6.099315     5.575342   
std      42.300571    24.284752    9981.264932     1.382997     1.112799   
min      20.000000    21.000000    1300.000000     1.000000     1.000000   
25%      20.000000    59.000000    7553.500000     5.000000     5.000000   
50%      50.000000    69.000000    9478.500000     6.000000     5.000000   
75%      70.000000    80.000000   11601.500000     7.000000     6.000000   
max     190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  \
count  1460.000000   1460.000000  1452.000000  1460.000000  1460.000000   
mean   1971.267808   1984.865753   103.685262   443.639726    46.549315   
std      30.202904     20.645407   181.066207   456.098091   161.319273   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000   
25%    1954.000000   1967.000000     0.000000     0.000000     0.000000   
50%    1973.000000   1994.000000     0.000000   383.500000     0.000000   
75%    2000.000000   2004.000000   166.000000   712.250000     0.000000   
max    2010.000000   2010.000000  1600.000000  5644.000000  1474.000000   

           ...         WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count      ...        1460.000000  1460.000000    1460.000000  1460.000000   
mean       ...          94.244521    46.660274      21.954110     3.409589   
std        ...         125.338794    66.256028      61.119149    29.317331   
min        ...           0.000000     0.000000       0.000000     0.000000   
25%        ...           0.000000     0.000000       0.000000     0.000000   
50%        ...           0.000000    25.000000       0.000000     0.000000   
75%        ...         168.000000    68.000000       0.000000     0.000000   
max        ...         857.000000   547.000000     552.000000   508.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  \
count  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000   
mean     15.060959     2.758904     43.489041     6.321918  2007.815753   
std      55.757415    40.177307    496.123024     2.703626     1.328095   
min       0.000000     0.000000      0.000000     1.000000  2006.000000   
25%       0.000000     0.000000      0.000000     5.000000  2007.000000   
50%       0.000000     0.000000      0.000000     6.000000  2008.000000   
75%       0.000000     0.000000      0.000000     8.000000  2009.000000   
max     480.000000   738.000000  15500.000000    12.000000  2010.000000   

           SalePrice  
count    1460.000000  
mean   180921.195890  
std     79442.502883  
min     34900.000000  
25%    129975.000000  
50%    163000.000000  
75%    214000.000000  
max    755000.000000  

[8 rows x 37 columns]

In [4]:
df.isnull().sum()

MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
BsmtQual           37
                 ... 
BedroomAbvGr        0
KitchenAbvGr        0
KitchenQual         0
TotRmsAbvGrd        0
Functional          0
Fireplaces          0
FireplaceQu       690
GarageType         81
GarageYrBlt        81
GarageFinish       81
GarageCars          0
GarageArea          0
GarageQual         81
GarageCond         81
PavedDrive

二、缺失值处理

1、把缺失值太多的数据直接去掉
原因：一共1400多条数据，就确实1300多条，无法进行分析。

In [5]:
df = df.drop(['PoolQC','Fence','MiscFeature','Alley','FireplaceQu'],axis=1)

2、缺失值处理：利用逻辑回归的方式进行填补缺失值LotFrontage

In [6]:
df = pd.get_dummies(df)
df_clean = df.dropna()
df_clean_X = df_clean.drop(columns=['SalePrice','LotFrontage','MasVnrArea','GarageYrBlt'])
df_clean_y = df_clean.SalePrice

In [7]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import Lasso

rfr = RandomForestRegressor()
lasso = Lasso()

In [8]:
from sklearn.model_selection import cross_val_score
rfr_cv = cross_val_score(rfr,df_clean_X,df_clean_y,scoring='neg_mean_squared_error',cv=5)
rfr_cv.mean()

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: Future

-1101388096.2149286

In [9]:
lasso_cv = cross_val_score(lasso,df_clean_X,df_clean_y,scoring='neg_mean_squared_error',cv=5)
lasso_cv.mean()

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase t

-1512796210.335291

In [10]:
from sklearn.model_selection import GridSearchCV
param_rfr = {'n_estimators':[10,30,100],
             'max_depth':list(range(2,11,2)),
             'min_samples_split':list(range(2,5,1)),
             'min_samples_leaf':list(range(1,4,1))}
# 使用格搜索查找最优超参数，优化目标是MSE
rfr_gs = GridSearchCV(rfr,param_rfr,scoring='neg_mean_squared_error',cv=5,verbose=1,
                      n_jobs=3).fit(df_clean_X,df_clean_y)
# 查看最优分数、最优超参数
rfr_gs.best_score_,rfr_gs.best_params_

[Parallel(n_jobs=3)]: Using backend LokyBackend with 3 concurrent workers.


Fitting 5 folds for each of 135 candidates, totalling 675 fits


[Parallel(n_jobs=3)]: Done  44 tasks      | elapsed:    3.4s
[Parallel(n_jobs=3)]: Done 194 tasks      | elapsed:   13.8s
[Parallel(n_jobs=3)]: Done 444 tasks      | elapsed:   45.4s
[Parallel(n_jobs=3)]: Done 675 out of 675 | elapsed:  1.5min finished
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\model_selection\_search.py:841: DeprecationWarning: The default of the `iid` parameter will change from True to False in version 0.22 and will be removed in 0.24. This will change numeric results when test-set sizes are unequal.
  DeprecationWarning)


(-1010340630.8254817,
 {'max_depth': 10,
  'min_samples_leaf': 1,
  'min_samples_split': 3,
  'n_estimators': 100})

In [11]:
# 使用格搜索查找Lasso最优超参数，因为Lasso超参数很少，因此没有单独设置需搜索的超参数字典。
lasso_gs = GridSearchCV(lasso,{'alpha':[1,0.1,0.01]},scoring='neg_mean_squared_error',
                        cv=5).fit(df_clean_X,df_clean_y)
# 查看分数和超参数，比较两个模型分数，选择RF作为Age缺失值的预测函数
lasso_gs.best_score_,lasso_gs.best_params_

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase t

(-1512024876.970385, {'alpha': 1})

In [12]:
df_clean_na = df.loc[df.LotFrontage.isna(),:]
df_clean_na_X = df_clean_na.drop(columns=['SalePrice','LotFrontage','MasVnrArea','GarageYrBlt'])
# 预测Age缺失值
df_clean_na_y = rfr_gs.best_estimator_.predict(df_clean_na_X)
# 保存为DataFrame格式（为了将index和数据结合起来，方便转化为dict）
df_clean_na_y = pd.DataFrame(df_clean_na_y,index=df_clean_na_X.index,columns=['LotFrontage'])
df = df.fillna(dict(df_clean_na_y))
df.isna().sum()

MSSubClass                0
LotFrontage               0
LotArea                   0
OverallQual               0
OverallCond               0
YearBuilt                 0
YearRemodAdd              0
MasVnrArea                8
BsmtFinSF1                0
BsmtFinSF2                0
BsmtUnfSF                 0
TotalBsmtSF               0
1stFlrSF                  0
2ndFlrSF                  0
LowQualFinSF              0
GrLivArea                 0
BsmtFullBath              0
BsmtHalfBath              0
FullBath                  0
HalfBath                  0
BedroomAbvGr              0
KitchenAbvGr              0
TotRmsAbvGrd              0
Fireplaces                0
GarageYrBlt              81
GarageCars                0
GarageArea                0
WoodDeckSF                0
OpenPorchSF               0
EnclosedPorch             0
                         ..
GarageFinish_RFn          0
GarageFinish_Unf          0
GarageQual_Ex             0
GarageQual_Fa             0
GarageQual_Gd       

3、缺失值处理：利用逻辑回归的方式进行填补缺失值GarageYrBlt

In [13]:
mean_garageyrblt = df['GarageYrBlt'].mean()
tmp = df['GarageYrBlt'].copy()
tmp[df.GarageYrBlt.isnull()] = mean_garageyrblt
df['GarageYrBlt_fill'] = tmp
del tmp
df = df.drop(['GarageYrBlt'],axis=1)

4、缺失值处理：将8个MasVnrArea缺失值赋值为0。

In [14]:
df.MasVnrArea= [x if str(x) != str('nan') else 0 for x in df.MasVnrArea]

In [15]:
df.isnull().sum().sum()

0

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 1 to 1460
Columns: 271 entries, MSSubClass to GarageYrBlt_fill
dtypes: float64(3), int64(34), uint8(234)
memory usage: 767.1 KB


In [17]:
df.columns

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       ...
       'SaleType_New', 'SaleType_Oth', 'SaleType_WD', 'SaleCondition_Abnorml',
       'SaleCondition_AdjLand', 'SaleCondition_Alloca', 'SaleCondition_Family',
       'SaleCondition_Normal', 'SaleCondition_Partial', 'GarageYrBlt_fill'],
      dtype='object', length=271)

In [18]:
df.sample(5)

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
Id                                                                            
313          190         65.0     7800            5            7       1939   
618           45         59.0     7227            6            6       1954   
621           30         45.0     8248            3            3       1914   
902           20         64.0     8712            5            7       1957   
1158         120         34.0     5001            7            5       2007   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2        ...         \
Id                                                            ...          
313           1950         0.0         507           0        ...          
618           1954         0.0           0           0        ...          
621           1950         0.0          41           0        ...          
902           2000         0.0         860           0        ...          
1158          2008       166.0         904           0        ...          

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
Id                                                                     
313              0             0            1                      0   
618              0             0            1                      0   
621              0             0            1                      0   
902              0             0            1                      0   
1158             0             0            1                      0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
Id                                                                        
313                       0                     0                     0   
618                       0                     0                     0   
621                       0                     0                     0   
902                       0                     0                     0   
1158                      0                     0                     0   

      SaleCondition_Normal  SaleCondition_Partial  GarageYrBlt_fill  
Id                                                                   
313                      1                      0       1939.000000  
618                      1                      0       1962.000000  
621                      1                      0       1978.506164  
902                      1                      0       1968.000000  
1158                     1                      0       2008.000000  

[5 rows x 271 columns]

三、进行预测、分析

1、首先进行逻辑回归进行预测：由于数据的汉字较多，我们需要哑变量

In [19]:
#df_dummies = pd.get_dummies(df, columns =['MSSubClass', 'LotFrontage','MSZoning','LotArea','Street','LotShape','LandContour','Utilities','LotConfig','LandSlope','Neighborhood','Condition1','Condition2','BldgType','HouseStyle','OverallQual','OverallCond','YearBuilt','YearRemodAdd','RoofStyl','RoofMatl','Exterior1st','Exterior2nd','MasVnrType','ExterQual','ExterCond','Foundation','BsmtQual','BsmtCond','BsmtExposure','BsmtFinType1','BsmtFinSF1','BsmtFinType2','BsmtFinSF2','BsmtUnfSF','TotalBsmtSF','Heating','HeatingQC','CentralAir','Electrical','1stFlrSF','2ndFlrSF','LowQualFinSF','GrLivArea','BsmtFullBath','BsmtHalfBath','FullBath','HalfBath','BedroomAbvGr','KitchenAbvGr','KitchenQual','TotRmsAbvGrd','Functional','Fireplaces','GarageType','GarageFinish','GarageCars','GarageArea','GarageQual','GarageCond','PavedDrive','WoodDeckSF','OpenPorch','EnclosedPorch','3SsnPorch','ScreenPorch','PoolArea','MiscVal','MoSold','YrSold','SaleType','SaleCondition','SalePrice','MasVnrArea_fill','GarageYrBlt_fill'])
df.columns

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       ...
       'SaleType_New', 'SaleType_Oth', 'SaleType_WD', 'SaleCondition_Abnorml',
       'SaleCondition_AdjLand', 'SaleCondition_Alloca', 'SaleCondition_Family',
       'SaleCondition_Normal', 'SaleCondition_Partial', 'GarageYrBlt_fill'],
      dtype='object', length=271)

In [20]:
X = df.loc[:,df.columns!='SalePrice']
y = df.loc[:,df.columns=='SalePrice']
from sklearn.model_selection import train_test_split

X_tr,X_ts,y_tr,y_ts = train_test_split(X,y,test_size=0.2)
X_tr.shape,y_tr.shape

((1168, 270), (1168, 1))

In [21]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()
model.fit(X_tr,y_tr)
model1=LogisticRegression(class_weight='balanced')
model1.fit(X_tr,y_tr)

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\svm\base.py:922: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  "the number of iterations.", ConvergenceWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWa

LogisticRegression(C=1.0, class_weight='balanced', dual=False,
          fit_intercept=True, intercept_scaling=1, max_iter=100,
          multi_class='warn', n_jobs=None, penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False)

In [22]:
y_pred = model.predict(X_ts)
pd.DataFrame(y_pred)[0].value_counts()

312500    6
139000    5
153000    4
108000    4
181000    4
142500    4
129500    3
254000    3
140000    3
208500    3
118400    3
225000    3
159000    3
90000     3
105000    3
84900     3
196000    3
135000    3
185000    3
154000    3
165000    3
115000    3
116000    3
269790    3
170000    3
261500    2
107500    2
176500    2
134000    2
153337    2
         ..
310000    1
142600    1
109900    1
118500    1
126175    1
150750    1
345000    1
199900    1
206900    1
125000    1
252678    1
309000    1
263435    1
133900    1
244000    1
136000    1
93500     1
325624    1
224500    1
222000    1
188000    1
153900    1
262280    1
137000    1
158500    1
180000    1
149500    1
173500    1
138000    1
179200    1
Name: 0, Length: 195, dtype: int64

In [23]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_ts,y_pred)

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [24]:
from sklearn.metrics import classification_report

print(classification_report(y_ts,y_pred))

              precision    recall  f1-score   support

       37900       0.00      0.00      0.00         1
       40000       0.00      0.00      0.00         0
       55993       0.00      0.00      0.00         0
       60000       0.00      0.00      0.00         2
       66500       0.00      0.00      0.00         0
       75500       0.00      0.00      0.00         1
       80000       0.00      0.00      0.00         1
       80500       0.00      0.00      0.00         1
       81000       0.00      0.00      0.00         0
       82500       0.00      0.00      0.00         0
       83000       0.00      0.00      0.00         2
       84500       0.00      0.00      0.00         2
       84900       0.00      0.00      0.00         0
       85000       0.00      0.00      0.00         0
       85400       0.00      0.00      0.00         1
       85500       0.00      0.00      0.00         0
       87000       0.00      0.00      0.00         1
       88000       0.00    

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1145: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


In [25]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
scores.mean()

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\utils\validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
C:\Users\DELL\

0.007794389054052985

结果分析：交叉验证结果接近于0，说明逻辑回归模型不好。

2、我们用RF、GBDT模型与逻辑回归模型进行对比

In [26]:
# 模型选择RF、GBDT
from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier

rfc = RandomForestClassifier()
gbc = GradientBoostingClassifier()

In [27]:
# 使用交叉验证计算RF分类器的准确率
from sklearn.model_selection import cross_val_score
rfc_cv = cross_val_score(rfc,X,y.values.ravel(),scoring='accuracy',cv=5)
rfc_cv.mean()

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
C:\Users\DELL\Anaconda3\lib\site-

0.0116104448275545

随机森林模型的精度比逻辑回归精度高一些

In [28]:
gbc_cv = cross_val_score(gbc,X,y.values.ravel(),scoring='accuracy',cv=5)
gbc_cv.mean()

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


0.009188099371180147

gbc模型没有随机森林模型的精度高

由于模型预测精度太低，我们没必要再进行格搜索

In [ ]:
# 使用格搜索寻找GBDT最优模型的超参数范围
param_gbc = {'min_samples_split':list(range(2,5,1)),
             'min_samples_leaf':list(range(1,5,1)),
             'max_depth':list(range(2,5,1))}
# 优化标准是准确率accuracy
gbc_gs = GridSearchCV(gbc,param_gbc,scoring='accuracy',cv=5).fit(X_tr,y_tr.values.ravel())
gbc_gs.best_score_,gbc_gs.best_params_

C:\Users\DELL\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:652: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=5.
  % (min_groups, self.n_splits)), Warning)


In [ ]:
# 设置使用格搜索寻找RF最优模型的超参数范围
param_rfc = {'min_samples_split':list(range(2,5,1)),
             'n_estimators':[10,100],
             'min_samples_leaf':list(range(1,5,1)),
             'max_depth':list(range(2,5,1))}
# 优化标准是准确率accuracy
rfc_gs = GridSearchCV(rfc,param_rfc,scoring='accuracy',cv=5).fit(X_tr,y_tr.values.ravel())
rfc_gs.best_score_,rfc_gs.best_params_

In [ ]:
# 设置使用格搜索寻找逻辑回归最优模型的超参数范围
param_logr = {'penalty':['l1','l2'],
              'C':[1e4,1e5]}
# 优化标准是准确率accuracy
logr_gs = GridSearchCV(logr,param_logr,scoring='accuracy',cv=5).fit(X_tr,y_tr.values.ravel())
logr_gs.best_score_,logr_gs.best_params_

总结：这三种方法计算精度都不高，因此没有必要继续进行格搜索优化模型，而计算精度不高的原因如下分析：
①数据缺失值过多，没有非常好的方法来填补缺失值，采用逻辑回归方式进行填补缺失值的方式之后的精度仍然不高；
②由于变量X过多，模型出现“过拟合”现象，造成精度不高；但是，不知道该移除哪一个变量；
③数据总量1460条，总量较少。